# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [1]:
import pandas as pd

temperatures = pd.Series(
    [72, 75, 68, 71, 73],
    index=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
)

print(f"The temperature series:\n{temperatures}")
print(f"\nData type:\n{temperatures.dtype}")

The temperature series:
Monday       72
Tuesday      75
Wednesday    68
Thursday     71
Friday       73
dtype: int64

Data type:
int64


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [2]:
products = pd.DataFrame({
    'product': ['Widget', 'Gadget', 'Doohickey'],
    'price': [19.99, 24.99, 12.49],
    'quantity': [100, 75, 150]
})

print(f"Dataframe:\n{products}")
print(f"\nShape: {products.shape}")

Dataframe:
     product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150

Shape: (3, 3)


#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [3]:
price = products['price']
print(f"Price column as a series:\n{price}")

prod_quant = products[['product', 'quantity']]
print(f"\nProduct and Quantity as a dataFrame:\n{prod_quant}")

Price column as a series:
0    19.99
1    24.99
2    12.49
Name: price, dtype: float64

Product and Quantity as a dataFrame:
     product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

##### Your Answers

#### 1
When a DataFrame is created from a dictionary, the keys become the column names and the values become column values.

#### 2
When I'm extracting data for one column by name (`df['price']`), `Pandas` gives me that column as a series. But, when I use `df[['price']]` structure for data extraction, I'm actually giving a list of column names. Even if the list has just one name, `Pandas` finds that I have asked for a subset of columns, hence gives me a DataFrame.”

#### 3
Custom index like task 1a would be helpful when the index itself carries meaning and will help me locate, extract, or read data more easily. For example, from the result of 1a, I can now extract the temperature for a specific day just by using its name without having to look for numeric index of that day.

### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [48]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('../data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)
df_raw = pd.read_csv(csv_path)

In [49]:
df_raw.head()

ride_id  rideable_type               started_at  \
0  67BB74BD7667BAB7  electric_bike  2024-09-30 23:12:01.622   
1  5AF1AC3BA86ED58C  electric_bike  2024-09-30 23:19:25.409   
2  7961DD2FC1280CDC   classic_bike  2024-09-30 23:32:24.672   
3  2E16892DEEF4CC19   classic_bike  2024-09-30 23:42:11.207   
4  AAF0220F819BEE01  electric_bike  2024-09-30 23:49:25.380   

                  ended_at         start_station_name start_station_id  \
0  2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave           bdd4c3   
1  2024-10-01 00:42:09.933                        NaN              NaN   
2  2024-10-01 00:23:18.647     St. Clair St & Erie St           9c619a   
3  2024-10-01 00:10:16.831  Ashland Ave & Chicago Ave           72a04d   
4  2024-10-01 00:06:27.476          900 W Harrison St           11da85   

           end_station_name end_station_id  start_lat  start_lng    end_lat  \
0                       NaN            NaN  42.012342 -87.688243  41.970000   
1    Benson Ave & Church St         a10cf0  42.070000 -87.730000  42.048214   
2  LaSalle St & Illinois St         fbd1ad  41.894345 -87.622798  41.890762   
3    Loomis St & Archer Ave         896337  41.895954 -87.667728  41.841633   
4         900 W Harrison St         11da85  41.874754 -87.649807  41.874754   

     end_lng member_casual  
0 -87.650000        casual  
1 -87.683485        casual  
2 -87.631697        member  
3 -87.657435        casual  
4 -87.649807        member

In [47]:
print("ndim:", df_raw.ndim)
print("\nshape:", df_raw.shape)
print(f"\ndata types: \n{df_raw.dtypes}")

ndim: 2

shape: (100000, 13)

data types: 
ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object


In [7]:
df_raw.nunique()

ride_id               100000
rideable_type              2
started_at             99998
ended_at               99987
start_station_name       825
start_station_id         825
end_station_name         827
end_station_id           827
start_lat               4595
start_lng               4586
end_lat                 1113
end_lng                 1103
member_casual              2
dtype: int64

#### Columns representing dates or categories
dates: `started_at`, `ended_at`

categories: `rideable_type`, `start_station_name`, `start_station_id`, `end_station_name`, `end_station_id`, `member_casual`

In [8]:
print("Memory usage by column:")
print(df_raw.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

Memory usage by column:
Index                     132
ride_id               7300000
rideable_type         6950493
started_at            8000000
ended_at              8000000
start_station_name    7549653
start_station_id      5978313
end_station_name      7546619
end_station_id        5974035
start_lat              800000
start_lng              800000
end_lat                800000
end_lng                800000
member_casual         6300000
dtype: int64

Total memory usage:
63.70 MB


#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [9]:
rides = pd.read_csv(
    csv_path,
    dtype={
        'rideable_type': 'category',
        'start_station_name': 'category',
        'start_station_id': 'category',
        'end_station_name': 'category',
        'end_station_id': 'category',
        'member_casual': 'category',
    },
    parse_dates=['started_at', 'ended_at']
)

In [10]:
rides.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  object        
 1   rideable_type       100000 non-null  category      
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   category      
 5   start_station_id    89623 non-null   category      
 6   end_station_name    89485 non-null   category      
 7   end_station_id      89485 non-null   category      
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  category      
dtypes: category(6), datetime64[ns]

In [11]:
print("Memory usage by column:")
print(rides.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{rides.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

Memory usage by column:
Index                     132
ride_id               7300000
rideable_type          100247
started_at             800000
ended_at               800000
start_station_name     300185
start_station_id       285039
end_station_name       300381
end_station_id         285165
start_lat              800000
start_lng              800000
end_lat                800000
end_lng                800000
member_casual          100126
dtype: int64

Total memory usage:
12.85 MB


#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [12]:
start_min = rides['started_at'].min()
start_max = rides['started_at'].max()
print(f"Start date range: {start_min} to {start_max}")

Start date range: 2024-09-30 23:12:01.622000 to 2024-10-31 23:54:02.851000


In [13]:
missing_vals = rides.isna().sum()
print(f"Missing values per column: \n{missing_vals}")

Missing values per column: 
ride_id                   0
rideable_type             0
started_at                0
ended_at                  0
start_station_name    10377
start_station_id      10377
end_station_name      10515
end_station_id        10515
start_lat                 0
start_lng                 0
end_lat                  87
end_lng                  87
member_casual             0
dtype: int64


In [14]:
df_missing = missing_vals.to_frame(name='count')
df_missing['percentage'] = (df_missing['count'] / len(rides)) * 100

print(f"Missing values with percentage: \n{df_missing}")

Missing values with percentage: 
                    count  percentage
ride_id                 0       0.000
rideable_type           0       0.000
started_at              0       0.000
ended_at                0       0.000
start_station_name  10377      10.377
start_station_id    10377      10.377
end_station_name    10515      10.515
end_station_id      10515      10.515
start_lat               0       0.000
start_lng               0       0.000
end_lat                87       0.087
end_lng                87       0.087
member_casual           0       0.000


#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [15]:
rides['trip_duration_min'] = (
    (rides["ended_at"] - rides["started_at"]).dt.total_seconds() / 60
)

In [16]:
rides["trip_duration_min"].describe()[["min", "max", "mean"]]

min        0.006533
max     1499.949717
mean      16.144576
Name: trip_duration_min, dtype: float64

In [17]:
rides[["started_at", "ended_at", "trip_duration_min"]].head()

started_at                ended_at  trip_duration_min
0 2024-09-30 23:12:01.622 2024-10-01 00:20:00.674          67.984200
1 2024-09-30 23:19:25.409 2024-10-01 00:42:09.933          82.742067
2 2024-09-30 23:32:24.672 2024-10-01 00:23:18.647          50.899583
3 2024-09-30 23:42:11.207 2024-10-01 00:10:16.831          28.093733
4 2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          17.034933

In [18]:
rides = rides.set_index('started_at')

In [20]:
print(f"After setting index: \n{rides.index}")
print(f"\nFirst few rows: \n{rides.head(5)}")

After setting index: 
DatetimeIndex(['2024-09-30 23:12:01.622000', '2024-09-30 23:19:25.409000',
               '2024-09-30 23:32:24.672000', '2024-09-30 23:42:11.207000',
               '2024-09-30 23:49:25.380000', '2024-09-30 23:49:40.016000',
               '2024-10-01 00:00:53.414000', '2024-10-01 00:05:44.954000',
               '2024-10-01 00:06:12.035000', '2024-10-01 00:10:03.646000',
               ...
               '2024-10-31 23:36:04.200000', '2024-10-31 23:36:34.956000',
               '2024-10-31 23:36:49.500000', '2024-10-31 23:38:20.262000',
               '2024-10-31 23:44:03.832000', '2024-10-31 23:44:23.211000',
               '2024-10-31 23:44:45.948000', '2024-10-31 23:50:31.160000',
               '2024-10-31 23:53:02.355000', '2024-10-31 23:54:02.851000'],
              dtype='datetime64[ns]', name='started_at', length=100000, freq=None)

First few rows: 
                                  ride_id  rideable_type  \
started_at                                     

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

#### 1
Pandas assigned `object` types to both `started_at` and `member_casual` in task 2a. Now, this constitutes problem because we don't know what is the actualy data type in those columns. Also, for datetime set as an object, we cannot do any operations such as, subtracting, resampling etc., on it. Setting `member_casual` as an object takes up more memory even though it has only two values, for a dataset large enough, this can slow down the actions. 

#### 2
Git commit IDs are derived from the commit’s entire content and context. It is more like a fingerprint for a particular commit than a serial number, as it includes its parents hash. I don't think the station IDs here aren’t completely cryptographic hashes like git commit IDs. Rather, they work as identifiers assigned to each stations so every ride can point to their respective station. I have also found some IDs are straight up numbers instead of crypotographic expressions and missing values assigned to some observations in the dataset. They’re maybe not derived from the station’s full characteristic, but might be based on the stations name and/or its location.

#### 3
Method chaining is the way of calling one method right after another on the same object so the output of one becomes the input to the next. `df.isna()` assesses each value in the dataset on each column to check whether there is a missing value or not and then assigns eithe True/False to that position. Here in `Pandas`, the boolean values are treated as 1 for True and 0 for False. Then, using the `.sum()` method adds up the number of Trues in each column and gives us the total count of missing values each column has.

#### 4
If about 10% of the station name or ID columns are missing but the location coordinates (lat/long) columns are almost complete, a possible explanation could be that the rider recorded their stating and ending coordinates, but the system couldn’t match these to station IDs or names. In that case, I think it would be better to not flag those row as bad data and remove them. Obviously we cannot use those rows for say a station-based analysis and interpretation, but for other analyses we can use the data in those rows. Whenever we are doing a staion-based analysis we can just filter the missing value rows and keep them outside of the scope of the analysis. We can also use the coordinates (since almost ~0% are missing) to find a way to generate a station name or ID.

#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

In 2a.3, the dataset was loaded with `Pandas` defaults which had multiple object data-types. These object types take more space in the memory. But, in 2b.3, we explicitly told `Pandas` which columns were categories and which were dates, so `Pandas` can store the category values once in a dictionary and just reference them. For this reason why the memory usage dropped in the latter case.  
The numbers we've seen by using `df.memory_usage(deep=True)` are higher and usually more accurate than what we've found from `df.info()`. This is because `.info()` returns a quick summary of the dataset and often underestimates the Python-object overhead for object columns. Now, if we actually care about if a data file can be handles by this memory size, we should use `df.memory_usage(deep=True).sum()`.


*Calculations for the dataset size:*  
Total RAM = 16 GB  
OS and other stuff = 30% of 16 GB = (0.30 × 16) = 4.8 GB  
Remaining memory = (16 − 4.8) GB = 11.2 GB  
If working with pandas needs 5–10× the dataset size:  
Optimistic (5×) = (11.2 GB / 5) = 2.24 GB dataset  
Conservative (10×) = (11.2 GB / 10) = 1.12 GB dataset  
So on a 16 GB machine I’d tell stay around 1–1.5 GB input data if I want things to be reliable and fast.

### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [75]:
mask = (rides['member_casual'] == "member") & (rides['rideable_type'] == "electric_bike")
member_electric_bike = rides[mask]

count_member_electric = len(member_electric_bike)
print(f"Trips taken by members using electric bikes: {count_member_electric}")

Trips taken by members using electric bikes: 33121


In [30]:
percent_member_electric = (count_member_electric / len(rides)) * 100
print(f"% trips taken by members using electric bikes: {percent_member_electric}")

% trips taken by members using electric bikes: 33.121


In [76]:
stations_mask = rides['start_station_name'].isin([
    "Streeter Dr & Grand Ave",
    "DuSable Lake Shore Dr & Monroe St",
    "Kingsbury St & Kinzie St"
])
trips_selected = rides[stations_mask]

print(f"No. of trips from selected stations: {len(trips_selected)}")

No. of trips from selected stations: 2702


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [77]:
rides['is_weekend'] = rides.index.dayofweek.isin([5, 6])

In [41]:
# verification
rides['is_weekend'].sample(10)

started_at
2024-10-30 14:40:10.527    False
2024-10-15 10:59:18.668    False
2024-10-01 16:10:01.960    False
2024-10-01 20:30:06.203    False
2024-10-18 08:11:37.939    False
2024-10-01 20:59:46.747    False
2024-10-07 06:58:51.821    False
2024-10-29 08:27:31.662    False
2024-10-09 10:29:33.858    False
2024-10-13 16:37:32.824     True
Name: is_weekend, dtype: bool

In [78]:
rides['time_of_day'] = "Midday"

start_hour = rides.index.hour

morning_mask = start_hour.isin([7, 8, 9])
rides.loc[morning_mask, 'time_of_day'] = "Morning Rush"

evening_mask = start_hour.isin([16, 17, 18])
rides.loc[evening_mask, 'time_of_day'] = "Evening Rush"

In [79]:
# verification
rides['time_of_day'].sample(10)

started_at
2024-10-26 10:43:32.674          Midday
2024-10-18 17:19:31.199    Evening Rush
2024-10-06 18:21:37.538    Evening Rush
2024-10-13 14:11:11.128          Midday
2024-10-20 12:03:03.003          Midday
2024-10-12 13:36:08.626          Midday
2024-10-09 18:40:47.425    Evening Rush
2024-10-26 10:54:53.272          Midday
2024-10-15 08:25:43.686    Morning Rush
2024-10-14 08:16:53.259    Morning Rush
Name: time_of_day, dtype: object

In [80]:
rides['time_of_day'].value_counts()

time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64

#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [56]:
filtered_df = rides.query(
    "member_casual == 'casual' and "
    "is_weekend == True and "
    "trip_duration_min > 20 and "
    "rideable_type == 'electric_bike'"
)

In [59]:
print(f"No. of trips matching the criteria: {len(filtered_df)}")

No. of trips matching the criteria: 1501


In [63]:
percent = (len(filtered_df) / len(rides)) * 100
print(f"% of trips meeting criteria: {percent:.2f}")

% of trips meeting criteria: 1.50


In [67]:
avg_duration = filtered_df["trip_duration_min"].mean()
print(f"Average duration of selected trips: {avg_duration:.2f} min")

Average duration of selected trips: 40.37 min


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [70]:
first_10 = rides.iloc[0:10][["member_casual", 
                             "rideable_type", 
                             "trip_duration_min"]]
print(first_10)

                        member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-09-30 23:12:01.622        casual  electric_bike          67.984200
2024-09-30 23:19:25.409        casual  electric_bike          82.742067
2024-09-30 23:32:24.672        member   classic_bike          50.899583
2024-09-30 23:42:11.207        casual   classic_bike          28.093733
2024-09-30 23:49:25.380        member  electric_bike          17.034933
2024-09-30 23:49:40.016        member  electric_bike          13.009367
2024-10-01 00:00:53.414        member   classic_bike           2.598817
2024-10-01 00:05:44.954        member  electric_bike           0.013433
2024-10-01 00:06:12.035        member  electric_bike          10.472933
2024-10-01 00:10:03.646        member   classic_bike           7.825683


In [73]:
oct_trips = rides.loc["2024-10-15":"2024-10-17"][["member_casual", 
                                                  "rideable_type", 
                                                  "trip_duration_min"]]
print(oct_trips)

                        member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-10-15 00:00:12.781        casual  electric_bike           2.804233
2024-10-15 00:01:20.517        member  electric_bike          11.330867
2024-10-15 00:05:24.811        member  electric_bike           1.868850
2024-10-15 00:05:52.984        member  electric_bike           2.705083
2024-10-15 00:06:18.819        member  electric_bike           1.600867
...                               ...            ...                ...
2024-10-17 23:45:48.739        member  electric_bike          14.703100
2024-10-17 23:47:35.040        member  electric_bike          13.049867
2024-10-17 23:55:34.112        member   classic_bike           3.795400
2024-10-17 23:56:14.464        member  electric_bike          11.937217
2024-10-17 23:59:38.103        casual   classic_bike           5.266433

[7235 rows x 3 columns]


In [74]:
print(f"Trips from 2024-10-15 to 2024-10-17: {len(oct_trips)}")

Trips from 2024-10-15 to 2024-10-17: 7235


#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

#### 1
Using `.isin()` is better when there's a long list of values, because then I can just pass the total list at once into the method, and `Pandas` can handle that efficiently. On the other hand, writing `(col == 'A') | (col == 'B') | ...` gets really messy, redundant, easy to make typos, and harder to maintain if the list changes.

#### 2
In Task 3b, we set everything to "Midday" first so that every row has some value, then we only overwrite the rows that match rush-hour times. If we didn’t set a default value to rows, some rows wouldn’t get any value. Similarly, if we had assigned the categories in a wrong order, we could have overwritten one label with another.

#### 3
`query()` is better when the logic is long and we want to have a readable filter (e.g., "member and weekend and duration > 20"). It enhances the readability and understandings of the filtering logic. Boolean indexing is better when we need variables in our condition. Using this, we can build masks step by step, reuse them, debug them, etc. Also it’s what we need when conditions might be used in loops, functions, etc.

#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

It’s better to use `[]` only for selecting columns and use `.loc[]` or `.iloc[]` for rows because bracket notation in pandas isn’t consistent for rows. For example, `df['col']` always means that get this specific column from the dataset, but `df[0:5]` only work for rows and creates a slice of the datset. So, `df[3]` won’t give us the third row, instead it will look for a column named 3, which may or may not exist. That mix of behaviors is easy to get tangled, especially for newbies. In contrast, `.loc[]` always means `by label` and `.iloc[]` always means `by position`, so what we are trying to do becomes clear and explicit. That is why it is better to use these instead using bracket notation for everything.

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [85]:
hour_counts = rides.index.hour.value_counts()
peak_hour = hour_counts.idxmax()
peak_hour_count = hour_counts.max()

print(f"The peak hour is {peak_hour}th hour")
print(f"{peak_hour_count} trips in the peak hour")

The peak hour is 17th hour
10574 trips in the peak hour


In [86]:
day_counts = rides.index.day_name().value_counts()
busiest_day = day_counts.idxmax()
busiest_day_count = day_counts.max()

print(f"The busiest day of the week is {busiest_day}")
print(f"{busiest_day_count} trips on {busiest_day}")

The busiest day of the week is Wednesday
16513 trips on Wednesday


In [113]:
days_order = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
day_counts_ordered = day_counts.reindex(days_order)
print(day_counts_ordered)

started_at
Monday       11531
Tuesday      14970
Wednesday    16513
Thursday     16080
Friday       13691
Saturday     14427
Sunday       12788
Name: count, dtype: int64


#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [94]:
avg_by_rider = rides.groupby("member_casual", 
                             observed=False)["trip_duration_min"].mean()
print(avg_by_rider)

member_casual
casual    23.978046
member    11.984493
Name: trip_duration_min, dtype: float64


In [95]:
avg_by_bike = rides.groupby("rideable_type",
                            observed=False)["trip_duration_min"].mean()
print(avg_by_bike)

rideable_type
classic_bike     20.337410
electric_bike    12.033618
Name: trip_duration_min, dtype: float64


In [101]:
counts_groupby = rides.groupby("member_casual", observed=False).size()
counts_vc = rides["member_casual"].value_counts()

print(counts_groupby)
print(counts_vc)

member_casual
casual    34686
member    65314
dtype: int64
member_casual
member    65314
casual    34686
Name: count, dtype: int64


#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [102]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Task 4c code here...
# use the variable `output_file` as the filename for step 4
# do not modify this setup code

In [103]:
weekend_electric = rides[
    (rides["is_weekend"] == True) &
    (rides["rideable_type"] == "electric_bike")
]

In [104]:
sampled = weekend_electric.iloc[:1000].reset_index()

In [105]:
cols_to_keep = ["started_at", "ended_at", "member_casual",
                "trip_duration_min", "time_of_day"]

sampled[cols_to_keep].to_csv(output_file, index=False)

In [109]:
total_weekend_electric = len(weekend_electric)
print(f"{total_weekend_electric} total weekend electric bike trips")

13026 total weekend electric bike trips


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

#### 1
`.groupby()` follows the "split–apply–combine" idea. First, it splits the dataframe into small groups based on the values in one column (e.g. all "member" rows together, all "casual" rows together). Then it applies some operation to each group separately. Finally, it combines the results into a new dataframe where each group has their corresponding result. So it’s a way to summarize, describe, or do some operation per group instead of for the whole dataset.

#### 2
`.value_counts()` is the fast, convenient way of counting how many times each value appears in one column. For example, this method is helpful for our data for counting how many members vs. casual. On the other hand, `.groupby().size()` is more general in behavior; we can group by column/s and then keep chaining other operations i.e., method chaining.  
If we needed counts by two columns (say, rider type and bike type), `.value_counts()` on one column won’t be enough, but `.groupby(["member_casual", "rideable_type"]).size()` would work and produce the result.

#### 3
We used `reset_index()` because the timestamp in our dataset was set previously as the index. Before exporing the `.csv` file of the sampled data we wanted `started_at` to appear as an actual column in the CSV. If we had exported the data while the` started_at` was still as the index and used `index=False`, the index wouldn’t be written at all. So, we would lose the datetime information in the exported file. Resetting index moves the `started_at` back into the columns, then we can safely do `index=False` and keep the timestamps.

#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [110]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

# Filter for interesting pattern: Long trips (>15 min) during rush hours 
# from lakefront stations, sorted by duration
lakefront_rush = (rides
    .loc[(rides.index.hour.isin([7, 8, 9, 16, 17, 18]))]
    .loc[(rides['start_station_name'].str.contains('Lake Shore|Lakefront', 
                                                    case=False, 
                                                    na=False))]
    .loc[rides['trip_duration_min'] > 15]
    .sort_values('trip_duration_min', ascending=False)
    .head(1000)
    .reset_index()
    [['started_at', 'ended_at', 'start_station_name', 'end_station_name',
      'member_casual', 'rideable_type', 'trip_duration_min']]
)

print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare file sizes
csv_size = os.path.getsize(csv_file) / 1024  # Convert to KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Compare load times
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Check data type preservation
# Note: CSV load without parse_dates loses datetime types
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)

print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)

Found 310 long rush-hour trips from lakefront stations

CSV file size: 40.57 KB
Pickle file size: 55.17 KB
Size difference: 14.59 KB

Load time comparison:
CSV:
2.15 ms ± 199 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

Pickle:
832 μs ± 10.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Data types from CSV (without parse_dates):
started_at             object
ended_at               object
start_station_name     object
end_station_name       object
member_casual          object
rideable_type          object
trip_duration_min     float64
dtype: object

Data types from Pickle:
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name          category
end_station_name            category
member_casual               category
rideable_type               category
trip_duration_min            float64
dtype: object


After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


##### Answer

#### 1
Putting the whole chain in parentheses and putting each method on its own line makes it way easier to read because then we can scan it top-to-bottom in an explicit chain. Debugging also becomes easier if this style used. We can temporarily comment out one line in the middle for checking where the issue is and then subsequently come up with proper solution to the problem. It also makes diffs in git cleaner, if I add or change one step, only that line shows up in the diff instead of one giant line. The downside of heavy chaining is that it can hide messy logic that would be clearer if it were split it into 2–3 steps. Also, if something went wrong in the middle of the chain, it would take some time to point out the real location of the issue.

#### 2
When I loaded the `.csv` file earlier without using `parse_dates=`, `Pandas` treated the datetime columns like as objects instead of assigning them their real type; similar thing happened to the categorical columns, they were also declared as objects. But, the `pickle` version kept the original dtypes for the columns and did not convert them to an `object` data-type. This become important because with the `pickled` version, I can immediately do time-based filtering, .dt.hour, groupby operations. Moreover, I don’t have to re-specify dtypes every time you load the file (which is a slight relief, I guess!).

#### 3
The `pickle` version would be great for my own workflow. It is faster to load, smaller in size, and preserves all the data-types. It is also perfect for iterative analysis in the same environment. `.csv` is better when I need portability and transparency meaning that any tool can open it, it’s human-readable, and it’s safer for sharing or long-term storage.

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

#### 1
The big shift is that with `NumPy` we think in terms of positions (i.e., row 0, col 1) on a purely numeric array, but with `Pandas` we think in terms of labels and columns with meaning. A dataframe isn’t just numbers, each column can have a name, a different type, etc., and `Pandas` line things up by these labels. It means that I can select by column name, work with dates, and handle missing data. So the mindset changes from "arrays of numbers" to "tables of labeled data" (I've borrowed this line from the web because it sounds nice and appropriate one-liner!!).

#### 2
Working with the real CSV was a bit messier than hw2b. Here, columns had different types (strings, dates, categories), some values were missing. Also timestamps were loaded as object that had to be parsed to revert it back to original data type, similar thing with the categorical values too. With the NumPy arrays, everything was already clean, numeric, same shape, and I just performed operations on them.

`Pandas` understand columns with names, mixed dtypes, missing data, and time indexes. It provides ways of handling missing data, assigning appropriate data types (if necessary), filtering data by labels, group by categories, resample, and export to CSV or Excel spreadsheets. These characteristics make `Pandas` a great framework for data analysis.

#### 3
Well, although I had some notions about the stuffs that we've used in this homework, I have learnt about how to assign appropriate data types, `groupby` actions, how to re-index, how to use `query`, and lastly the `pickle` version of loading data. These will help me a lot in working with my own data. Also, i have learnt about how these things work by writing answers to the reflection questions which has helped me to get a better understanding of what I'm actually doing.  
I would say a solid 7.5 on using `Pandas` for my own projects. I don't want repeat the same talk again and again, but with the variety of operations that `Pandas` offer, it will be much more helpful platform to load, handle, and analyze real-life datasets. My current score (a whopping 7.5 out 10!!!) will eventually get better with more and more experience with messy, messier, and messiest datasets (hopefully). (xD)

#### 4
Time spent: 8 hours (coding: 2.5 hrs; debugging: 2.5 hrs; writing: 3 hrs) (mentally I'd say it felt like a 20-hour stretch) (:/)

Most helpful part of the assignment: Well, it was all enjoyable and the guides were really helpful, they explicitly told what to do and gave hints too. I specially enjoyed using git commands (felt like a hacker! xD).

One specific improvement suggestion: So far so good. But the amount of writing can be reduced a bit (or make it long enough so that it feels a 48 hours stretch and paralyzes my back, just joking!!)

P.S.-1: Yaay, the assingment is done!!  
P.S.-2: Excuse the silly things I have written, I'm just trying to make you laugh!!  
![It's a cat!](https://i.kym-cdn.com/photos/images/original/001/384/531/8ed.jpg)